File: backend/c/domain/value/account_type.c
#include "../include/account_type.h"
#include <string.h>
#include <strings.h>

const char *account_type_as_str(AccountType t) {
    switch (t) {
        case ACCOUNT_TYPE_SAVINGS: return "Savings";
        case ACCOUNT_TYPE_CREDIT:  return "Credit";
        case ACCOUNT_TYPE_REGULAR: return "Regular";
        default:                   return "Unknown";
    }
}

bool account_type_from_str(const char *s, AccountType *out, DomainError *err) {
    if (!s || !out) {
        if (err) *err = domain_error_validation("account type string is null");
        return false;
    }

    if (!strcasecmp(s, "Savings")) {
        *out = ACCOUNT_TYPE_SAVINGS;
        if (err) *err = domain_error_ok();
        return true;
    }
    if (!strcasecmp(s, "Credit")) {
        *out = ACCOUNT_TYPE_CREDIT;
        if (err) *err = domain_error_ok();
        return true;
    }
    if (!strcasecmp(s, "Regular")) {
        *out = ACCOUNT_TYPE_REGULAR;
        if (err) *err = domain_error_ok();
        return true;
    }

    if (err) *err = domain_error_validation("Invalid account type");
    return false;
}

const AccountType *account_type_all(size_t *out_count) {
    static const AccountType ALL[ACCOUNT_TYPE_COUNT] = {
        ACCOUNT_TYPE_SAVINGS,
        ACCOUNT_TYPE_CREDIT,
        ACCOUNT_TYPE_REGULAR
    };
    if (out_count) *out_count = ACCOUNT_TYPE_COUNT;
    return ALL;
}

File: backend/rust/src/domain/value/account_type.rs
use core::fmt;
use core::str::FromStr;

use crate::domain::errors::DomainError;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum AccountType {
    Savings,
    Credit,
    Regular,
}

impl AccountType {
    pub fn as_str(&self) -> &str {
        match self {
            AccountType::Savings => "Savings",
            AccountType::Credit => "Credit",
            AccountType::Regular => "Regular",
        }
    }

    pub fn all() -> &'static [AccountType] {
        &[AccountType::Savings, AccountType::Credit, AccountType::Regular]
    }
}

impl fmt::Display for AccountType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for AccountType {
    type Err = DomainError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.trim();
        if s.eq_ignore_ascii_case("Savings") { return Ok(AccountType::Savings); }
        if s.eq_ignore_ascii_case("Credit")  { return Ok(AccountType::Credit); }
        if s.eq_ignore_ascii_case("Regular") { return Ok(AccountType::Regular); }
        Err(DomainError::validation("Invalid account type"))
    }

}

impl From<AccountType> for String {
    fn from(value: AccountType) -> Self {
        value.as_str().to_string()
    }
}
